### PR TITLE
util/limit: always check ctx, add unit test

### PR DESCRIPTION
**Pull request by dt**

### CODE REVIEW

**Comment by dt** [comment]
I ran this test with a much higher number of requests -- taking about 30s per run, as well as with and without `-race`, but no sign of a hang yet.

File: pkg/util/limit/limiter_test.go [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+	"runtime"
+	"sync/atomic"
+	"testing"
+
+	"golang.org/x/sync/errgroup"
+
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+)
+
+func TestConcurrentRequestLimiter(t *testing.T) {
```
Comment (mjibson) [inline pkg/util/limit/limiter_test.go]: Do we need to run this test on every CI build? Seems excessive. Maybe skip it if t.Short is on?

Comment (dt) [reply pkg/util/limit/limiter_test.go]: I thought about that, or lowering runs if t.Short was on, but it currently runs in .2s so it didn't seem that bad?

**Review by madelynnblue — APPROVED** [review]

**Review by petermattis — APPROVED** [review]
:lgtm:

Btw, @nvanbenschoten has a PR for a resizable semaphore: https://github.com/cockroachdb/cockroach/pull/29622. The implementation of that semaphore started as a fork of `golang.org/x/sync/semaphore`. Something to investigate. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/31594#-:-LPBUgOj0lce1ipbDwgs:b9jj8tv)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale)

**Pull request merged**